e2e/e2e_test.go

```diff
@@ -66,97 +66,92 @@ func simpleReadWrite(errChan chan error, outChan chan string, conn io.ReadWriter
 	}
 }
 
-func assertE2ECommunication(ctx context.Context, clientConfig, serverConfig *dtls.Config, serverPort int, t *testing.T) {
-	var (
-		messageRecvCount uint64 // Counter to make sure both sides got a message
-		clientMutex      sync.Mutex
-		clientConn       net.Conn
-		serverMutex      sync.Mutex
-		serverConn       net.Conn
-		serverListener   net.Listener
-		serverReady      = make(chan struct{})
-		errChan          = make(chan error)
-		clientChan       = make(chan string)
-		serverChan       = make(chan string)
-	)
-
-	// DTLS Client
-	go func() {
-		select {
-		case <-serverReady:
-			// OK
-		case <-time.After(time.Second):
-			errChan <- errors.New("waiting on serverReady err: timeout")
-		}
-
-		clientMutex.Lock()
-		defer clientMutex.Unlock()
+type comm struct {
+	ctx                        context.Context
+	clientConfig, serverConfig *dtls.Config
+	serverPort                 int
+	messageRecvCount           *uint64 // Counter to make sure both sides got a message
+	clientMutex                *sync.Mutex
+	clientConn                 net.Conn
+	serverMutex                *sync.Mutex
+	serverConn                 net.Conn
+	serverListener             net.Listener
+	serverReady                chan struct{}
+	errChan                    chan error
+	clientChan                 chan string
+	serverChan                 chan string
+	useClientOpenSSL           bool
+	useServerOpenSSL           bool
+}
 
-		var err error
-		clientConn, err = dtls.DialWithContext(ctx, "udp",
-			&net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: serverPort},
-			clientConfig,
-		)
-		if err != nil {
-			errChan <- err
-			return
-		}
+func newComm(ctx context.Context, clientConfig, serverConfig *dtls.Config, serverPort int, mode string) *comm {
+	messageRecvCount := uint64(0)
+	c := &comm{
+		ctx:              ctx,
+		clientConfig:     clientConfig,
+		serverConfig:     serverConfig,
+		serverPort:       serverPort,
+		messageRecvCount: &messageRecvCount,
+		clientMutex:      &sync.Mutex{},
+		serverMutex:      &sync.Mutex{},
+		serverReady:      make(chan struct{}),
+		errChan:          make(chan error),
+		clientChan:       make(chan string),
+		serverChan:       make(chan string),
+	}
+	switch mode {
+	case "openssl_client":
+		c.useClientOpenSSL = true
+	case "openssl_server":
+		c.useServerOpenSSL = true
+	}
+	return c
+}
 
-		simpleReadWrite(errChan, clientChan, clientConn, &messageRecvCount)
-	}()
+func (c *comm) assert(t *testing.T) {
+	// DTLS Client
+	if c.useClientOpenSSL {
+		go c.clientOpenSSL()
```

Comment:
I think it's better to keep ordinary tests being able to work without openssl.

Reply:
I agree :+1: